ts_src/taproot.ts

```diff
@@ -0,0 +1,256 @@
+ */
+const EVEN_Y_COORD_PREFIX = new Uint8Array([0x02]);
+const INITIAL_TAPSCRIPT_VERSION = new Uint8Array([0xc0]);
+
+const TAGS = [
+  'TapLeaf',
+  'TapBranch',
+  'TapTweak',
+  'KeyAgg list',
+  'KeyAgg coefficient',
+] as const;
+type TaggedHashPrefix = typeof TAGS[number];
+/** An object mapping tags to their tagged hash prefix of [SHA256(tag) | SHA256(tag)] */
+const TAGGED_HASH_PREFIXES = Object.fromEntries(
+  TAGS.map(tag => {
+    const tagHash = bcrypto.sha256(Buffer.from(tag));
+    return [tag, Buffer.concat([tagHash, tagHash])];
+  }),
+) as { [k in TaggedHashPrefix]: Buffer };
+
+function taggedHash(prefix: TaggedHashPrefix, data: Buffer): Buffer {
+  return bcrypto.sha256(Buffer.concat([TAGGED_HASH_PREFIXES[prefix], data]));
+}
+
+/**
+ * Trims the leading 02/03 byte from an ECDSA pub key to get a 32 byte schnorr
+ * pub key with x-only coordinates.
+ * @param pubkey A 33 byte pubkey representing an EC point
+ * @returns a 32 byte x-only coordinate
+ */
+export function trimFirstByte(pubkey: Buffer): Buffer {
+  assert.strictEqual(pubkey.length, 33);
+  return pubkey.slice(1, 33);
+}
+
+/**
+ * Aggregates a list of public keys into a single MuSig2* public key
+ * according to the MuSig2 paper.
+ * @param pubkeys The list of pub keys to aggregate
+ * @returns a 32 byte Buffer representing the aggregate key
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
+  // TODO: Consider enforcing key uniqueness.
+  assert(
+    pubkeys.length > 1,
+    'at least two pubkeys are required for musig key aggregation',
+  );
+
+  // Trim the 0x02/0x03 leading byte from each key and sort in ascending order
+  // to convert to 32 byte x-coordinates with implicit even Y coordinates.
+  const trimmedPubkeys = pubkeys.map(trimFirstByte).sort(Buffer.compare);
+
+  // In MuSig all signers contribute key material to a single signing key,
+  // using the equation
+  //
+  //     P = sum_i µ_i * P_i
+  //
+  // where `P_i` is the public key of the `i`th signer and `µ_i` is a so-called
+  // _MuSig coefficient_ computed according to the following equation
+  //
+  // L = H(P_1 || P_2 || ... || P_n)
+  // µ_i = H(L || P_i)
+
+  const L = taggedHash('KeyAgg list', Buffer.concat(trimmedPubkeys));
+
+  const tweakedPubkeys: Buffer[] = trimmedPubkeys.map(
+    (trimmedPubkey, index) => {
+      const pubkey = Buffer.concat([EVEN_Y_COORD_PREFIX, trimmedPubkey]);
+
+      if (index === 1) {
+        // The second unique key in the pubkey list gets the constant KeyAgg
+        // coefficient 1 which saves an exponentiation. See the MuSig2*
+        // appendix in the MuSig2 paper for details.
+        return pubkey;
+      }
+
+      const c = taggedHash(
+        'KeyAgg coefficient',
+        Buffer.concat([L, trimmedPubkey]),
+      );
+      return ecc.pointMultiply(pubkey, c);
+    },
+  );
+  const aggregatePubkey = tweakedPubkeys
+    .slice(1)
+    .reduce((prev, curr) => ecc.pointAdd(prev, curr), tweakedPubkeys[0]);
```

Comment:
> initialValue Optional
>    A value to which previousValue is initialized the first time the callback is called. If initialValue is specified, that also causes currentValue to be initialized to the first value in the array. If initialValue is not specified, previousValue is initialized to the first value in the array, and currentValue is initialized to the second value in the array.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce
This can be simplified to
```suggestion
const aggregatePubkey = tweakedPubkeys.reduce((prev, cur) => ecc.pointAdd(prev, curr));
```

Reply:
Yep that checks out, the slice is superfluous.

Reply:
Yeah, I suggested the slice, but I'm a bit rusty still, sorry! :)